canister/src/api/get_block_headers.rs

```diff
@@ -1,17 +1,139 @@
+
+    if request.start_height > chain_height {
+        return Err(GetBlockHeadersError::StartHeightDoesNotExist {
+            requested: request.start_height,
+            chain_height,
+        });
+    }
+
     if let Some(end_height) = request.end_height {
         if end_height < request.start_height {
             return Err(GetBlockHeadersError::StartHeightLagerThanEndHeight {
                 start_height: request.start_height,
                 end_height,
             });
         }
+
+        if end_height > chain_height {
+            return Err(GetBlockHeadersError::EndHeightDoesNotExist {
+                requested: end_height,
+                chain_height,
+            });
+        }
+        // If `end_height` is provided then it should be the
+        // end of effective height range.
+        Ok((request.start_height, end_height))
+    } else {
+        // If `end_height`` is not provided then the end of effective
+        // range should be the last block of the chain.
+        Ok((request.start_height, chain_height))
+    }
+}
+
+fn get_block_headers_internal(
+    request: &GetBlockHeadersRequest,
+) -> Result<(GetBlockHeadersResponse, Stats), GetBlockHeadersError> {
+    let (start_height, end_height) =
+        verify_requested_height_range_and_return_effective_range(request)?;
+    // The last stable block is located in the unstable_blocks, hence the height of the
+    // last block located in stable_blocks if it exists is `s.stable_height() - 1`.
+    let height_of_last_block_in_stable_blocks = with_state(|s| {
+        let stable_height = s.stable_height();
+        if stable_height > 0 {
+            Some(stable_height - 1)
+        } else {
+            None
+        }
+    });
+
+    let mut stats: Stats = Stats::default();
+
+    // Build block headers vec.
+    let ins_start = performance_counter();
+
+    let mut vec_headers = vec![];
+
+    // Add requested block headers located in stable_blocks.
+    if let Some(height_of_last_block_in_stable_blocks) = height_of_last_block_in_stable_blocks {
+        if start_height <= height_of_last_block_in_stable_blocks {
+            let end_range_in_stable_blocks =
+                std::cmp::min(height_of_last_block_in_stable_blocks, end_height);
+
+            vec_headers = with_state(|s| {
+                let block_heights = &s.stable_block_headers.block_heights;
+                let block_headers = &s.stable_block_headers.block_headers;
+                block_heights
+                    .range(start_height..=end_range_in_stable_blocks)
+                    .map(|(_, block_hash)| block_headers.get(&block_hash).unwrap().into())
+                    .collect()
+            });
+        }
     }
-    Ok(())
+
+    // How the last stable block is located in unstable_blocks, there will always
+    // be the block in unstable_blocks.
+    let height_of_first_block_in_unstable_blocks =
+        height_of_last_block_in_stable_blocks.map_or(0, |h| h + 1);
```

Comment:
I think this can be made much clearer to read:
```suggestion
    let height_of_first_block_in_unstable_blocks = state.stable_height();
```

Reply:
Done.